feat(idempotency): adding redis as idempotency backend (#2567)
awslabs/aws-lambda-powertools-python: docs/utilities/idempotency.md

* Select a subset of the event as the idempotency key using JMESPath expressions
* Set a time window in which records with the same payload should be considered duplicates
* Expires in-progress executions if the Lambda function times out halfway through

> * Support Amazon DynamoDB and Redis as persistence layer

Reviewer: ```suggestion
* Support Amazon DynamoDB and Redis as persistence layers
```

## Getting started

???+ note

>     This section uses DynamoDB as default idempotent persistence storage layer. If you are interested in using Redis as persistence storage layer, Check out the [Redis as persistence storage layer](#redis-as-persistent-storage-layer-provider) Section.

Reviewer: ```suggestion
    This section uses DynamoDB as the default idempotent persistence storage layer. If you are interested in using Redis as the persistence storage layer, check out the [Redis as persistence storage layer](#redis-as-persistent-storage-layer-provider) section.
```

Before getting started, you need to create a persistent storage layer where the idempotency utility can store its state - your lambda functions will need read and write access to it.

> We currently support Amazon DynamoDB and Redis as a storage layer. This example demonstrates how to create a table in DynamoDB. If you want to use Redis, please go to the section [RedisPersistenceLayer](#redispersistencelayer)

Reviewer: ```suggestion
We currently support Amazon DynamoDB and Redis as storage layers. The following example demonstrates how to create a table in DynamoDB. If you prefer to use Redis, please refer to the [RedisPersistenceLayer](#redispersistencelayer) section.
```

    As this happens outside the scope of your decorated function, you are not able to catch it if you're using the `idempotent` decorator on your Lambda handler.

### Persistence layers

#### DynamoDBPersistenceLayer

> This persistence layer is built-in, and you can either use an existing DynamoDB table or create a new one dedicated for idempotency state (recommended).

Reviewer: ```suggestion
This persistence layer is built-in, allowing you to use an existing DynamoDB table or create a new one dedicated to idempotency state (recommended).
```

=== "Customizing DynamoDBPersistenceLayer to suit your table structure"

    ```python hl_lines="7-15"
    --8<-- "examples/idempotency/src/customize_persistence_layer.py"
    ```

> When using DynamoDB as a persistence layer, you can alter the attribute names by passing these parameters when initializing the persistence layer:

Reviewer: ```suggestion
When using DynamoDB as the persistence layer, you can customize the attribute names by passing the following parameters during the initialization of the persistence layer:
```

| **status_attr**             |                    | `status`                             | Stores status of the lambda execution during and after invocation                                        |
| **data_attr**               |                    | `data`                               | Stores results of successfully executed Lambda handlers                                                  |
| **validation_key_attr**     |                    | `validation`                         | Hashed representation of the parts of the event used for validation                                      |
| **sort_key_attr**           |                    |                                      | Sort key of the table (if table is configured with a sort key).                                          |
| **static_pk_value**         |                    | `idempotency#{LAMBDA_FUNCTION_NAME}` | Static value to use as the partition key. Only used when **sort_key_attr** is set.                       |

#### RedisPersistenceLayer

> This persistence layer is built-in, and you can use an existing Redis service. For optimal performance and compatibility, we strongly advise using a Redis service version 7 or higher.

Reviewer: ```suggestion
This persistence layer is built-in, allowing you to use an existing Redis service. For optimal performance and compatibility, it is strongly recommended to use a Redis service version 7 or higher.
```

=== "Customizing RedisPersistenceLayer to suit your data structure"

    ```python hl_lines="14-20"
    --8<-- "examples/idempotency/src/customize_persistence_layer_redis.py"
    ```

> When using Redis as a persistence layer, you can alter the attribute names by passing these parameters when initializing the persistence layer:

Reviewer: ```suggestion
When using Redis as the persistence layer, you can customize the attribute names by providing the following parameters upon initialization of the persistence layer:
```

| Parameter                   | Required           | Default                              | Description                                                                                              |
| --------------------------- | ------------------ | ------------------------------------ | -------------------------------------------------------------------------------------------------------- |
| **in_progress_expiry_attr** |                    | `in_progress_expiration`             | Unix timestamp of when record expires while in progress (in case of the invocation times out)            |

> | **status_attr**             |                    | `status`                             | Stores status of the lambda execution during and after invocation                                        |

Reviewer: ```suggestion
| **status_attr**             |                    | `status`                             | Stores status of the Lambda execution during and after invocation                                        |
```

<i>Optional idempotency key</i>
</center>

## Redis as persistent storage layer provider

### Redis resources

> You need an existing Redis service before setting up Redis as persistent storage layer provider. You can also use Redis compatible services like [Amazon ElastiCache for Redis](https://aws.amazon.com/elasticache/redis/){target="_blank"} or [Amazon MemoryDB for Redis](https://aws.amazon.com/memorydb/){target="_blank"} as persistent storage layer provider.

Reviewer: ```suggestion
Before setting up Redis as the persistent storage layer provider, you must have an existing Redis service. Alternatively, you can use Redis compatible services such as [Amazon ElastiCache for Redis](https://aws.amazon.com/elasticache/redis/){target="_blank"} or [Amazon MemoryDB for Redis](https://aws.amazon.com/memorydb/){target="_blank"} as your persistent storage layer provider.

```

???+ tip "No existing Redis service?"

>     If you don't have an existing Redis service, we recommend using [DynamoDB](#dynamodbpersistencelayer) as persistent storage layer provider.

Reviewer: ```suggestion
    If you don't have an existing Redis service, we recommend using [DynamoDB](#dynamodbpersistencelayer) as the persistent storage layer provider.
```

=== "AWS CloudFormation example"

    ```yaml hl_lines="5"
    --8<-- "examples/idempotency/templates/cfn_redis_serverless.yaml"
    ```

    1. Replace the Security Group ID and Subnet ID to match your VPC settings.

### VPC Access

> Your Lambda Function must be able to reach the Redis endpoint before using it for idempotency persistent storage layer. In most cases you will need to [configure VPC access](https://docs.aws.amazon.com/lambda/latest/dg/configuration-vpc.html){target="_blank"} for your Lambda Function.

Reviewer: ```suggestion
Your Lambda Function must have network access to the Redis endpoint before using it as the idempotency persistent storage layer. In most cases, you will need to [configure VPC access](https://docs.aws.amazon.com/lambda/latest/dg/configuration-vpc.html){target="_blank"} for your Lambda Function.
```

> ???+ tip "Amazon ElastiCache/MemoryDB for Redis as persistent storage layer provider"
>     If you intend to use Amazon ElastiCache for Redis for idempotency persistent storage layer, you can also reference [This AWS Tutorial](https://docs.aws.amazon.com/lambda/latest/dg/services-elasticache-tutorial.html){target="_blank"}.
>     If you are using Amazon MemoryDB for Redis, reference [This AWS Tutorial](https://aws.amazon.com/blogs/database/access-amazon-memorydb-for-redis-from-aws-lambda/){target="_blank"} for only VPC setup part.

Reviewer: ```suggestion
???+ tip "Amazon ElastiCache/MemoryDB for Redis as persistent storage layer provider"
    If you plan to use Amazon ElastiCache for Redis as the idempotency persistent storage layer, you may find [this AWS tutorial](https://docs.aws.amazon.com/lambda/latest/dg/services-elasticache-tutorial.html){target="_blank"} helpful.
    For those using Amazon MemoryDB for Redis, refer to [this AWS tutorial](https://aws.amazon.com/blogs/database/access-amazon-memorydb-for-redis-from-aws-lambda/){target="_blank"} specifically for the VPC setup guidance.
```

> After VPC setup, you can follow the templates down below to setup Lambda functions with VPC internal subnet access.

Reviewer: ```suggestion
After completing the VPC setup, you can use the templates provided below to set up Lambda functions with access to VPC internal subnets.
```

=== "AWS Serverless Application Model (SAM) example"

    ```yaml hl_lines="9"
    --8<-- "examples/idempotency/templates/sam_redis_vpc.yaml"
    ```

    1. Replace the Security Group ID and Subnet ID to match your VPC settings.

### Configuring Redis persistence layer

> You can quickly start by initializing the `RedisCachePersistenceLayer` class and using it with the `idempotent` decorator on your lambda handler. Check out detailed example of `RedisCachePersistenceLayer` in [Persistence layers section](#redispersistencelayer).

Reviewer: ```suggestion
You can quickly get started by initializing the `RedisCachePersistenceLayer` class and applying the `idempotent` decorator to your Lambda handler. For a detailed example of using the `RedisCachePersistenceLayer`, refer to the [Persistence layers section](#redispersistencelayer).
```

=== "Use established Redis Client"
    ```python hl_lines="4 7 12-15 17 31"
    --8<-- "examples/idempotency/src/getting_started_with_idempotency_redis_client.py"
    ```

=== "Use Persistence Layer with Redis config variables"
    ```python hl_lines="4-8 10 24"
    --8<-- "examples/idempotency/src/getting_started_with_idempotency_redis_config.py"
    ```

=== "Sample event"

    ```json
    --8<-- "examples/idempotency/src/getting_started_with_idempotency_payload.json"
    ```

### Custom advanced settings

> For advanced settings, including SSL certificates and the ability to customize parameters such as a custom timeout, you can use the Redis client to accommodate these specific settings.

Reviewer: ```suggestion
For advanced configurations, such as setting up SSL certificates or customizing parameters like a custom timeout, you can utilize the Redis client to tailor these specific settings to your needs.
```

    --8<-- "examples/idempotency/tests/app_test_io_operations.py"
    ```

### Testing with Redis

> To test locally, You can either utilize [fakeredis-py](https://github.com/cunla/fakeredis-py) or check out the [MockRedis](https://github.com/aws-powertools/powertools-lambda-python/blob/ba6532a1c73e20fdaee88c5795fd40e978553e14/tests/functional/idempotency/persistence/test_redis_layer.py#L34-L66) Class we used in our test.

Reviewer: ```suggestion
To test locally, you can either utilize [fakeredis-py](https://github.com/cunla/fakeredis-py) for a simulated Redis environment or refer to the [MockRedis](https://github.com/aws-powertools/powertools-lambda-python/blob/ba6532a1c73e20fdaee88c5795fd40e978553e14/tests/functional/idempotency/persistence/test_redis_layer.py#L34-L66) class used in our tests to mock Redis operations.
```

=== "test_with_mock_redis.py"

    ```python hl_lines="2 3 29 31"
    --8<-- "examples/idempotency/tests/test_with_mock_redis.py"
    ```

=== "mock_redis.py"

    ```python
    --8<-- "examples/idempotency/tests/mock_redis.py"
    ```

> If you want to actually setup a Real Redis client for integration test, reference the code below

Reviewer: ```suggestion
If you want to set up a real Redis client for integration testing, you can reference the code provided below.
```
